In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sb

from sklearn.cluster import KMeans

In [2]:
df = pd.read_csv('avocado_clean.csv')

df = df.drop(columns=['Date'])

# K-means para dos clases, comprobar la relevancia de los datos para predecir el tipo de aguacate (columna type)

In [30]:
from sklearn.metrics.cluster import completeness_score, homogeneity_score, v_measure_score
from sklearn.preprocessing import LabelEncoder, StandardScaler, MinMaxScaler


kmeans = KMeans(n_clusters=2, random_state=0)

aux_df = pd.get_dummies(df, columns = ['region'],drop_first=True)


#Label encoding la columna de tipod de aguacate
le = LabelEncoder()
aux_df['type'] = le.fit_transform(aux_df['type'])
real_label = aux_df['type']

aux_df = aux_df.drop(columns= ['type'])

#Estandarizando variables
scaler = MinMaxScaler()
aux_scaled = scaler.fit_transform(aux_df)


kmeans.fit(aux_scaled)

kmeans_label = kmeans.labels_

print("Homogeneity:", homogeneity_score(real_label,kmeans_label))
print("Completeness:", completeness_score(real_label,kmeans_label))
print("V-measure:", v_measure_score(real_label,kmeans_label))


Homogeneity: 0.4248302673129039
Completeness: 0.4166786256143546
V-measure: 0.4207149642531975


In [26]:
df['type'].value_counts()

,count
type,
organic,8304
conventional,2646


Sorpendentemente, las métricas obtenidas indican que el k means no es una técnica útil para predecir el tipo de aguacate de cada fila. A lo mucho puede hacerlo para aproximadamente la mitad de datos del dataset, lo cual se observa en la métrica de completitud y homogeneidad. Esto es interesante, ya que, como se observó en otros modelos, la clasificación entre estos dos tipos se puede realizar hasta de forma visual, ya que muchas veces los datos de aguacates orgánicos tienden a estar cercanos al 0.

Esto puede ser debido al gran desbalance de ambas clases, hay 8304 aguacates orgánicos y 2646 convencionales.

# K-means para más de dos clases, comprobar la relevancia de los datos para predecir la region del aguacate (columna region)

El problema de desbalance de clases es significativamente menor para las regiones. Por lo que de obtener métricas bajas o deficientes indicarían que los datos no son útiles para las regiones de las que provienen los aguacates.

In [29]:
from sklearn.metrics.cluster import completeness_score, homogeneity_score, v_measure_score
from sklearn.preprocessing import LabelEncoder, StandardScaler, MinMaxScaler


kmeans = KMeans(n_clusters=len(df['region'].value_counts()), random_state=0)

aux_df = pd.get_dummies(df, columns = ['type'],drop_first=True)


#Label encoding la columna de tipod de aguacate
le = LabelEncoder()
aux_df['region'] = le.fit_transform(aux_df['region'])
real_label = aux_df['region']

aux_df = aux_df.drop(columns= ['region'])

#Estandarizando variables
scaler = MinMaxScaler()
aux_scaled = scaler.fit_transform(aux_df)


kmeans.fit(aux_scaled)

kmeans_label = kmeans.labels_


print("Homogeneity:", homogeneity_score(real_label,kmeans_label))
print("Completeness:", completeness_score(real_label,kmeans_label))
print("V-measure:", v_measure_score(real_label,kmeans_label))


Homogeneity: 0.33120276295584855
Completeness: 0.3841053659112549
V-measure: 0.3556977848342342


De nuevo, se obtienen resultados un tanto deficientes, por lo menos considerando la tarea de intentar clasificar los aguacates en las distintas regiones perfectamente. Sin embargo, considerando que son alrededor de 50 regiones, los resultados obtenidos muestran que los datos dan la suficiente información como para que un tercio de dichos datos sean clasificados correctamente entre todas las regiones. El gran número de regiones no indica que por lo menos el modelo no realiza clasificaciones en las 53 regiones completamente al azar.